### DISCUSSION TITLE
How to fill text in a rectangle shape without defiing font size?

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#3008] [Q&A]
How to fill text in a rectangle shape without defiing font size?

**Comment by JorjMcKie** [comment]
I recommend to use the new Page method [insert_htmlbox](https://pymupdf.readthedocs.io/en/latest/page.html#Page.insert_htmlbox).
It is similar to `Page.insert_textbox()` and `TextWriter.fill_textbox()` but very much more powerful.
You can not only write plain text, but also almost arbitrary mixtures of text with HTML / CSS language particles.

So - if you know elementary HTML / CSS syntax - you can change font, color, boldness etc. in one single string that you pass as a parameter.
This includes tables created via HTML tag `<table>` and even images (`<img>` tag).
Any missing font for Unicodes will automatically be pulled in (from Google Noto fonts).

And you can stop worrying about whether your content will fit:
If you use the default, the content (i.e. text, images etc.) will automatically be scaled down accordingly.

**Comment by nissansz** [comment]
convert_pdf_to_ppt(pdf, pptx)
  File "F:/pycharm2020.2/RapidStructure-0.0.0/OnnxDetRec_pdf_writeOcrtxtLine_html20240110.py", line 923, in convert_pdf_to_ppt
    page.insert_htmlbox(rect, text, css=None, scale_low=0, archive=None, rotate=0, oc=0, opacity=1, overlay=True)
AttributeError: 'Page' object has no attribute 'insert_htmlbox'


Any sample code to replace TextWriter?

  # tw.write_text(page, opacity=None, color=color, morph=(pos, mat), overlay=True, oc=0, render_mode=0)

     # tw = fitz.TextWriter(page.rect)
            # # tw.append((50, 100), text, font=font, small_caps=True)
            # # tw.append((int(left), int(bottom)), text,  fontsize=(int(bottom) - int(top))*0.7, font=font, small_caps=True)
            # tw.append((int(left), int(top)), text,  fontsize=(int(bottom) - int(top))*0.7, font=font, small_caps=True)
            # # tw.write_text(page)
            # tw.write_text(page, opacity=None, color=color, morph=None, overlay=True, oc=0, render_mode=0)
           
            rect = (int(left), int(top), int(right), int(bottom))
            # page.insert_htmlbox(rect, text, *, css=None, scale_low=0, archive=None, rotate=0, oc=0, opacity=1, overlay=True)
            page.insert_htmlbox(rect, text, css=None, scale_low=0, archive=None, rotate=0, oc=0, opacity=1, overlay=True)

**Reply by JorjMcKie** [reply]
Examples are [here](https://pymupdf.readthedocs.io/en/latest/recipes-text.html#how-to-fill-a-box-with-html-text).
To take influence on the appearance, mobilize your knowledge about HTML and CSS. This lets you achieve various font sizes, colors, bold, italic, font, text alignment and more.
The method correspondingly has no parameters like font size or color. But you can rotate the content (again: text and all other content) and set opacity plus overlay.

**Comment by nissansz** [comment]
but there is below error when try to use it.

page = doc[i]

   page.insert_htmlbox(rect, text, css="* {font-family: sans-serif;font-size:14px;}")
AttributeError: 'Page' object has no attribute 'insert_htmlbox'

**Reply by JorjMcKie** [reply]
Ah, forgot to mention that you must use `import fitz_new as fitz`, because the method only exists in the new rebased implementation.
Best you use this import statement to stay stable across our move to rebased in the next few weeks:
```python
try:
    import fitz_new as fitz
except ImportError:
    import fitz
...
```

**Comment by nissansz** [comment]
Still same error after import fitz_new 

    page.insert_htmlbox(rect, text, css="* {font-family: sans-serif;font-size:14px;}")
AttributeError: 'Page' object has no attribute 'insert_htmlbox'

**Reply by JorjMcKie** [reply]
your pymupdf version?

**Comment by nissansz** [comment]
Still same error after import fitz_new 

    page.insert_htmlbox(rect, text, css="* {font-family: sans-serif;font-size:14px;}")
AttributeError: 'Page' object has no attribute 'insert_htmlbox'

**Comment by nissansz** [comment]
PyMuPDF                      1.20.2
PyMuPDFb                     1.23.3
PyNaCl                       1.5.0
pyparsing                    3.0.9
PyPDF2                       1.23
pyperclip                    1.8.2
pypiwin32                    223
pyppeteer                    1.0.2
pyppmd                       0.18.3

**Reply by JorjMcKie** [reply]
this is a **_new_** method, available since 1.23.8

**Comment by nissansz** [comment]
[_ダイス_ molding - Google 搜索_overlay.pdf](https://github.com/pymupdf/PyMuPDF/files/13889939/_._.molding.-.Google._overlay.pdf)
can write, but text position seems wrong.



            rect = fitz.Rect((int(left), int(top), int(right), int(bottom)))
            # page.insert_htmlbox(rect, text, *, css=None, scale_low=0, archive=None, rotate=0, oc=0, opacity=1, overlay=True)
            # page.insert_htmlbox(rect, text, css=None, scale_low=0, archive=None, rotate=0, oc=0, opacity=1, overlay=True)
            page.insert_htmlbox(rect, text,  css="* {font-family: sans-serif;font-size:14px;}")

**Reply by JorjMcKie** [reply]
forgot to do `page.clean_contents()` before everything else?

**Comment by nissansz** [comment]
Thank you. Will play once I have time.

**Comment by nissansz** [comment]
1. Span background color did not appear with html code. Font color can appear.

 <td>Between<br>Gray and Purple<span style="background-color:aqua;color:red">text<span></td>
[testhtmlwrite.pdf](https://github.com/pymupdf/PyMuPDF/files/13896580/testhtmlwrite.pdf)

2. When I rotated image by 90 degrees etc. in original pdf by adobe, then save.
After that, when read by pymupdf, the image did not rotate, still the old image direction. 

3. When insert text with insert_htmlbox, the font size did not autoshape to evenly distribut text in the rect so that it can match original image distribution, there is large margin if there are many characters. Any parameter like morph for insert_htmlbox?
![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/39231e55-75f1-4ad6-bc4b-50f4c8ae24f1)

**Comment by nissansz** [comment]
when try to use this method, the result is not correct.
https://github.com/pymupdf/PyMuPDF/issues/259

   textwidth = fitz.get_text_length(text,  fontsize=fontsize)
            # pivot = fitz.Point(rect.x0, rect.y0 + rect.height / 2)  # middle of left side
            # pos = pivot
            mat = fitz.Matrix(textwidth/rect.width, 1.0)
            # mat = fitz.Matrix(rect.width/textwidth, 1.0)

            # # mat = fitz.Matrix(1/0.7, 1)  # 2 = double the width
            # tw.write_text(page, opacity=None, color=color, morph=(pos, mat), overlay=True, oc=0, render_mode=0)
            tw.write_text(page, opacity=None, color=None, morph=None, overlay=False, oc=0, render_mode=0)
            # result = page.insert_textbox(rect, text, fontsize=fontsize, color=(1, 0, 0), morph=(pos, mat))

![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/bee9f093-2f39-4471-806e-0b263585d77f)

**Comment by nissansz** [comment]
Is it possible to support htm canvas/javascript when using insert_htmlbox? Example effect?

It is blank when I try to insert text with below code.

'''
text = """
	<!DOCTYPE html>
<html>
<head>
    <title>Canvas Text Fill</title>
</head>
<body>
    <canvas id="canvas" width="400" height="200"></canvas>

    <script>
        var canvas = document.getElementById('canvas');
        var ctx = canvas.getContext('2d');

        var text = "Hello World";
        var areaWidth = 200; // 长方形区域宽度
        var areaHeight = 100; // 长方形区域高度
        var fontSize = 40; // 初始字体大小

        // 调整字体大小以填充满指定区域
        while (true) {
            ctx.font = fontSize + "px Arial";
            var textWidth = ctx.measureText(text).width;
            var textHeight = fontSize * 0.7; // 假设文字高度为字体大小的 70%

            if (textWidth <= areaWidth && textHeight <= areaHeight) {
                break;
            }

            fontSize--;
        }

        // 居中绘制文本
        var x = (canvas.width - textWidth) / 2;
        var y = (canvas.height + textHeight) / 2;
        ctx.fillText(text, x, y);
    </script>
</body>
</html>
"""

'''

**Comment by JorjMcKie** [comment]
> Is it possible to support htm canvas/javascript when using insert_htmlbox? Example effect?


No, javascript is not supported, only HTML (v4) and CSS (v2) syntax. To go into more detail, please read chapter 16 about Story of the free book [Mupdf Explored](https://mupdf.com/docs/mupdf-explored.html).

> Span background color did not appear with html code. Font color can appear.

`<span>` as an inline-level element cannot have a background color - see HTML / CSS documentation. Only block-level elements can. Experiment with `<div>` which is a block-level element, but has according layout consequences. Example:
```python
import fitz


rect = fitz.Rect(100, 100, 400, 300)

text = """Lorem ipsum dolor sit amet, consectetur adipisici elit, sed
    eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad
    minim veniam, quis nostrud exercitation <b>ullamco <i>laboris</i></b> 
    nisi ut aliquid ex ea commodi consequat. Quis aute iure 
    <div style="color: red;background-color: yellow;">reprehenderit</div> 
    in <span style="color: green;font-weight:bold;">voluptate</span> velit 
    esse cillum dolore eu fugiat nulla pariatur. Excepteur sint obcaecat 
    cupiditat non proident, sunt in culpa qui 
    <a href="https://www.artifex.com">officia</a> deserunt mollit anim id 
    est laborum."""

doc = fitz.Document()

page = doc.new_page()
page.insert_htmlbox(rect, text, css="* {font-family: sans-serif;font-size:14px;}")

doc.ez_save(__file__.replace(".py", ".pdf"))
```
Generates this:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/6e4fd8f3-a04d-476b-8cbc-fee73e72bc27)


> When I rotated image by 90 degrees etc. in original pdf by adobe, then save. After that, when read by pymupdf, the image did not rotate, still the old image direction.

Don't understand. This code
```python
import fitz

rect = fitz.Rect(100, 250, 300, 350)
text = """<img src="nur-ruhig.jpg">"""
doc = fitz.open()
page = doc.new_page()
page.insert_htmlbox(rect, text, archive=fitz.Archive("."), opacity=0.5, rotate=90)
page.draw_rect(rect, color=(1, 0, 0))
doc.save(__file__.replace(".py", ".pdf"))
```
Produces this page: 
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/13850cd2-3f56-4c39-aae8-64e557d59eca)
When I then retrieve the image information, I learn that it indeed is rotated:
```python
>>> page.get_images()
[(6, 0, 439, 501, 8, 'DeviceRGB', '', 'Img3', 'DCTDecode')]

>>> pprint(page.get_image_rects(6,transform=True))
[(Rect(101.0, 251.0, 212.84054565429688, 349.0),
  Matrix(0.0, -98.0, 111.84054565429688, 0.0, 101.0, 349.0))]
```

> Any parameter like morph for insert_htmlbox?

No there is not .. yet.

**Comment by nissansz** [comment]
[2.pdf](https://github.com/pymupdf/PyMuPDF/files/13899260/2.pdf)

Attached a file with Adobe OCR layer.
I want to get below info.
Textline box, font info, font name, alignment etc. so that I can imitate when adding my own result.

How to unhide the OCR layer from Adobe?
![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/56bdebbc-24a1-4c42-9a36-a08c48cce6b5)

**Reply by JorjMcKie** [reply]
Can we please have different Discussions posts for different questions?
Nobody will learn anything when looking at this thread - and will not even **_like_** to go into the details because it looks like higgledy-piggledy.

No there is no easy way to do that.

**Comment by nissansz** [comment]
Is there any external python method to calculate proper font size for html rectganle shape so as to realize autofit in pymupdf through insert_htmlbox?

**Reply by JorjMcKie** [reply]
No there is not. That would be against the purpose of the method, namely to not have to care about layout issues and be able to mix fonts, font sizes, tables and images and let the method do the rest.

**Comment by nissansz** [comment]
ok
